src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs

```diff
@@ -137,18 +121,50 @@ private async Task<bool> HandleHttpResponseAsync(HttpResponseMessage response)
 
             if (!obj.TryGetValue("KeyId", out var keyId) || keyId.Type != JTokenType.String)
             {
-                throw new ArgumentNullException("Missing required <KeyId> field.");
+                throw new AzureSignalRException("Missing required <KeyId> field.");
             }
             if (!obj.TryGetValue("AccessKey", out var key) || key.Type != JTokenType.String)
             {
-                throw new ArgumentNullException("Missing required <AccessKey> field.");
+                throw new AzureSignalRException("Missing required <AccessKey> field.");
             }
             Key = new Tuple<string, string>(keyId.ToString(), key.ToString());
 
-            _authorizeTcs.TrySetResult(true);
             return true;
         }
 
+        private async Task UpdateAccessKeyAsyncCore(IServerNameProvider provider, ILoggerFactory loggerFactory)
+        {
+            _timer.Start();
+
+            var logger = loggerFactory.CreateLogger<AadAccessKey>();
+
+            while (await _timer)
+            {
+                for (int i = 0; i < AuthorizeMaxRetryTimes; i++)
+                {
+                    var source = new CancellationTokenSource(AuthorizeTimeout);
+                    try
+                    {
+                        await AuthorizeAsync(provider.GetName(), source.Token);
+                        Log.SucceedAuthorizeAccessKey(logger, Endpoint);
+                        break;
+                    }
+                    catch (Exception e)
+                    {
+                        _isAuthorized = false;
```

Comment:
should this `false` set move outside of the retry logic, considering the case when the previous auth is succeeded?

Reply:
If the previous auth is succeeded, we will simply break the for...loop and there will be no retry afterward.